sdk/dotnet/Pulumi/Serialization/Serializer.cs

```diff
@@ -385,6 +385,29 @@ internal static Value CreateValue(object? value)
                 _ => throw new InvalidOperationException("Unsupported value when converting to protobuf: " + value.GetType().FullName),
             };
 
+        /// <summary>
+        /// Detects encoded `Unknown` values in objects that conform
+        /// to the grammar returned by `SerializeAsync`.
+        ///
+        /// This possibly needs to be revisited to detect `Unknown`
+        /// values before `SerializeAsync` converts them, in the more
+        /// generic Output representation.
+        /// </summary>
+        internal static bool ContainsUnknowns(object? value)
+            => value switch
+            {
+                null => false,
+                int _ => false,
+                double d => false,
+                bool b => false,
+                string s => s == Constants.UnknownValue,
+                ImmutableArray<object> list => list.Any(ContainsUnknowns),
+                ImmutableDictionary<string, object> dict =>
+                    dict.Values.Any(ContainsUnknowns) ||
+                    dict.Keys.Contains(Constants.UnknownValue),
+                _ => throw new InvalidOperationException("Unsupported value when converting to protobuf: " + value.GetType().FullName),
+            };
```

Comment:
You've heard Alan Perlis quote that "functional programmers know the Value of everything but the Cost of nothing"? This is very much the situation here :) I'll humor you and incorporate the changes but as you point it's a wash, say Pulumi program has 10,000 outputs that go through this,  10 * 1e-9 * 10000 = 1E-5 seconds savings, this is going to be noise for a Pulumi program like that. I'm very curious if someone runs programs that either have lots of outputs or very deep trees being passed around; understanding those scenarios would actually help construct some benchmarks to be mindful of this.

Reply:
I'd like to leave 'unsupportedValue' check in place though to remind us when the assumptions of the grammar break. If we introduce new kidns of values there that contain unknowns this will help find/fix the code.